src/fairseq2/gang.py

```diff
@@ -690,15 +693,63 @@ def __post_init__(self) -> None:
                     "Coordinator process of the root gang (i.e. `root.rank == 0`) must be rank 0 in all parallel gangs."
                 )
 
+    def __enter__(self) -> None:
+        _thread_local.current_gangs.append(self)
+
+    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+        _thread_local.current_gangs.pop()
+
     def close(self) -> None:
+        """Destroys all gangs."""
         self.root.close()
 
 
+_thread_local = threading.local()
+
+# Holds the stack of current thread-local gangs.
+_thread_local.current_gangs = []
+
+
+def maybe_get_current_gangs() -> Gangs | None:
+    """
+    Returns the current gangs to use for collective operations.
+
+    By default, this function returns ``None``. The current gangs of the calling
+    thread can be set by using :class:`Gangs` as a context manager:
+
+    .. code::
+
+        from fairseq2.gang import Gangs
+
+        gangs = Gangs(...)
+
+        with gangs:
+            current_gangs = maybe_get_current_gangs()
+
+            assert current_gangs is gangs
+
+        current_gangs = maybe_get_current_gangs()
+
+        assert current_gangs is None
+
+    Within fairseq2, this function is used by model factories to retrieve the
+    current gangs and shard the constructed models accordingly. The current gangs
+    are set internally by fairseq2 before calling the factories.
+
+    Note that the return value of this function is thread specific. Individual
+    threads may have their own set of current gangs.
+    """
+    if _thread_local.current_gangs:
+        return cast(Gangs, _thread_local.current_gangs[-1])
```

Comment:
Cool trick! Any specific reason we might expect to have nested context gang calls for this to be index -1 instead of 0?